This notebook contains the initial steps taken for data quality check and consistency of the 'Uber dataset'.

1. Checking shape and data types

2. Looking for missing values

3. Looking for missing data

4. Dealing with negative values

5. Changing data types in some columns

6. Locating outliers

In [1]:
# Importing Analitycal and Visual libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

Creating a path to import and work on data set

In [2]:
path = r'C:\Users\Paola\Desktop\Shaul\Data\CF\Data Immersion\Achievement 6\Uber October-2023 Basket Analysis'

In [3]:
path

'C:\\Users\\Paola\\Desktop\\Shaul\\Data\\CF\\Data Immersion\\Achievement 6\\Uber October-2023 Basket Analysis'

Importing Uber data set

In [4]:
df = pd.read_csv(os.path.join(path, '2-Data', 'Original Data', 'uber_data_wrangled.csv'), index_col = False)

Checking shape and columns

In [5]:
df.shape

(100000, 21)

In [6]:
df.columns

Index(['VendorID', 'pickup_date', 'pickup_time', 'dropff_date', 'dropoff_time',
       'passenger_count', 'trip_distance', 'pickup_longitude',
       'pickup_latitude', 'RatecodeID', 'store_and_fwd_flag',
       'dropoff_longitude', 'dropoff_latitude', 'payment_type', 'fare_amount',
       'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'improvement_surcharge', 'total_amount'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   VendorID               100000 non-null  int64  
 1   pickup_date            100000 non-null  object 
 2   pickup_time            100000 non-null  object 
 3   dropff_date            100000 non-null  object 
 4   dropoff_time           100000 non-null  object 
 5   passenger_count        100000 non-null  int64  
 6   trip_distance          100000 non-null  float64
 7   pickup_longitude       100000 non-null  float64
 8   pickup_latitude        100000 non-null  float64
 9   RatecodeID             100000 non-null  int64  
 10  store_and_fwd_flag     100000 non-null  object 
 11  dropoff_longitude      100000 non-null  float64
 12  dropoff_latitude       100000 non-null  float64
 13  payment_type           100000 non-null  int64  
 14  fare_amount            100000 non-nul

Addressing data quality, consistency and initial preparation

In [8]:
# Viewing the missing values
df.isnull().sum()

VendorID                 0
pickup_date              0
pickup_time              0
dropff_date              0
dropoff_time             0
passenger_count          0
trip_distance            0
pickup_longitude         0
pickup_latitude          0
RatecodeID               0
store_and_fwd_flag       0
dropoff_longitude        0
dropoff_latitude         0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
dtype: int64

No missing data.

In [9]:
# Finding duplicates
df_dups = df[df.duplicated()]

In [10]:
df_dups.shape

(0, 21)

No duplicates in this data set.

In [10]:
# Checking df for mixing data
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

No mixed data in this data set.

In [44]:
# This command has been repeated a few times, to check that replacement with mean value was succesful.
df.describe()

trip_distance  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  100000.000000     100000.000000    100000.000000      100000.000000   
mean        3.034270        -73.288983        40.375220         -73.312418   
std         3.846951          7.089652         3.901413           6.964171   
min         0.000000       -121.933327         0.000000        -121.933327   
25%         0.990000        -73.990959        40.738891         -73.990547   
50%         1.670000        -73.980202        40.755299         -73.978416   
75%         3.200000        -73.964203        40.769021         -73.962090   
max       184.400000          0.000000        41.204548           0.000000   

       dropoff_latitude    fare_amount          extra        mta_tax  \
count     100000.000000  100000.000000  100000.000000  100000.000000   
mean          40.388064      13.265219       0.101809       0.497658   
std            3.833974      11.674715       0.201996       0.034125   
min            0.000000       0.000000       0.000000       0.000000   
25%           40.738541       6.500000       0.000000       0.500000   
50%           40.755079       9.500000       0.000000       0.500000   
75%           40.767910      15.000000       0.000000       0.500000   
max           42.666893     819.500000       4.500000       0.500000   

          tip_amount   tolls_amount  improvement_surcharge   total_amount  
count  100000.000000  100000.000000          100000.000000  100000.000000  
mean        1.872670       0.367415               0.299916      16.406226  
std         2.618513       1.527758               0.005019      14.424392  
min         0.000000       0.000000               0.000000       0.000000  
25%         0.000000       0.000000               0.300000       8.300000  
50%         1.360000       0.000000               0.300000      11.800000  
75%         2.460000       0.000000               0.300000      18.300000  
max       125.880000      25.540000               0.300000     832.800000

There seems to be issues with some inconsistency in the entries. There are many outliers as well as negative amounts in the 'fare_amount' column. Also, some of the columns have numeric values, but they represent options, not values.

Dealing with the negative values.

In [40]:
# Checking the negative values. This and the following two commands have been repeated for the following columns:
# Columns 'fare_amount', 'extra', 'mat_tax', 'tip_amount', 'improvement_surcharge' and 'total_amount'.
df.loc[df['total_amount'] < 0]

Empty DataFrame
Columns: [VendorID, pickup_date, pickup_time, dropff_date, dropoff_time, passenger_count, trip_distance, pickup_longitude, pickup_latitude, RatecodeID, store_and_fwd_flag, dropoff_longitude, dropoff_latitude, payment_type, fare_amount, extra, mta_tax, tip_amount, tolls_amount, improvement_surcharge, total_amount]
Index: []

[0 rows x 21 columns]

The 'fare_amount' column had negative values in 70 rows. This does not make sense, as independently of the length of the trip, any form of payment would have meant a positive value. It is reasonable to ascribe this error to a typo mistake. As it is only 0.0007% of the data, I will use the mean value for imputation to solve this issue.

The 'extra' column had negative values in 19 rows. This does not make sense, as independently of the length of the trip, any form of extras would have meant a positive value. It is reasonable to ascribe this error to a typo mistake. As it is only 0.00019% of the data, I will use the mean value for imputation to solve this issue.

The 'mta_tax' column had negative values in 66 rows. This does not make sense, as independently of the length of the trip, any form of tax would have meant a positive value. It is reasonable to ascribe this error to a typo mistake. As it is only 0.00066% of the data, I will use the mean value for imputation to solve this issue.

The 'tip_amount' column had negative values in 4 rows. This does not make sense, as independently of the length of the trip, any form of tip would have meant a positive value. It is reasonable to ascribe this error to a typo mistake. As it is only 0.0001% of the data, I will use the mean value for imputation to solve this issue.

The 'improvement_surcharge' column had negative values in 70 rows. This does not make sense, as independently of the length of the trip, any form of payment would have meant a positive value. It is reasonable to ascribe this error to a typo mistake. As it is only 0.0007% of the data, I will use the mean value for imputation to solve this issue.

The 'improvement_surcharge' column had negative values in 70 rows. This does not make sense, as independently of the length of the trip, any form of payment would have meant a positive value. It is reasonable to ascribe this error to a typo mistake. As it is only 0.0007% of the data, I will use the mean value for imputation to solve this issue.

In [38]:
# Replacing  negative outliers with the mean. 
df.loc[df['total_amount'] < 0, 'total_amount'] = np.nan

In [39]:
# Negative outliers solved => same operation repeated for all columns just named above.
df['total_amount'].fillna((df['total_amount'].mean()), inplace=True)

Changing data types from numeric to string.

In [43]:
# Changing the following columns data types to string (text): VendorID, passenger_count, payment_type, RateCodeID. 
# Even though written in numbers, they represent text options, not numbers. This comand has been repeated for all the abovementioned columns.
df['payment_type'] = df['payment_type'].astype('str')

The numeric columns are: trip_distance, fare_amount, extra, mta_tax, tip_amount, tolls_amount, mprovement_surcharge, total_amount.
Pickup_time and dropoff_time alre also numeric, but currently their data is shown as object.

In [64]:
# Checking the outliers in the max values.
# This command has been repeated with different values, and with both columns 'trip distance' and 'fare amount', to get the report below.
df.loc[df['fare_amount'] > 300]

VendorID       pickup_date pickup_time       dropff_date dropoff_time  \
35136        2  10/03/2016 00:00    10:18:00  10/03/2016 00:00     18:32:00   
75653        1  01/03/2016 00:00    01:02:00  01/03/2016 00:00     03:54:00   
76891        2  10/03/2016 00:00    13:41:00  10/03/2016 00:00     13:42:00   
77168        2  10/03/2016 00:00    13:43:00  10/03/2016 00:00     13:43:00   

      passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
35136               2          51.27        -73.947281        40.783939   
75653               1         160.80        -73.905586        40.874123   
76891               1           0.00        -73.996964        40.742344   
77168               1           0.00        -73.996964        40.742344   

      RatecodeID  ... dropoff_longitude  dropoff_latitude  payment_type  \
35136          1  ...        -73.951141         40.785179             2   
75653          4  ...        -73.710709         42.666893             2   
76891          5  ...        -73.996964         40.742344             1   
77168          5  ...          0.000000          0.000000             1   

      fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
35136       314.5    0.0      0.5         0.0           0.0   
75653       819.5    0.5      0.5         0.0          12.0   
76891       400.0    0.0      0.0         0.0           0.0   
77168       400.0    0.0      0.0         0.0           0.0   

       improvement_surcharge  total_amount  
35136                    0.3         315.3  
75653                    0.3         832.8  
76891                    0.3         400.3  
77168                    0.3         400.3  

[4 rows x 21 columns]

From 100000 trips, there are:
Just 2 that exceed 100 miles (entries 75653 and 99665)
Another 2 that exceed 50 miles (entries 35136 and 94698)
Another 551 entries that exceed 20 miles
Another 2355 entries that exceed 15 miles. In total, 2910 entries.

There are 2921 trips that exceed $50 dollars. From those, 64 that exceed $100 dollars. From the latter, 26 that exceed $150 dollars. Lastly, from those 10 exceed $200 dollars. Only four trips exceed $300 dollars.

There seems to be no relation between length of trip and fare amount, as I can observe some relatively short trips costing very expensive, and some long trips costing less. Also, observing pick up and drop times it seems that fare amount is judged by time in the cab, not length of trip. I would go to my superior to check this.

In [65]:
# Exporting 'df' to csv
df.to_csv(os.path.join(path, '2-Data','Prepared Data', 'uber_checked.csv'))